File: api/models/cardio.py
"""
Cardiovascular disease analysis models.

This module defines the Pydantic models for cardiovascular disease analysis.
"""

from enum import Enum
from typing import Dict, List, Optional, Union

from pydantic import BaseModel, Field


class Gender(int, Enum):
    """Gender enumeration."""
    FEMALE = 1
    MALE = 2


class CholesterolLevel(int, Enum):
    """Cholesterol level enumeration."""
    NORMAL = 1
    ABOVE_NORMAL = 2
    WELL_ABOVE_NORMAL = 3


class GlucoseLevel(int, Enum):
    """Glucose level enumeration."""
    NORMAL = 1
    ABOVE_NORMAL = 2
    WELL_ABOVE_NORMAL = 3


class PatientData(BaseModel):
    """Patient data model."""
    age: int = Field(..., description="Age of the patient in years")
    gender: Gender = Field(..., description="Gender of the patient")
    height: float = Field(..., description="Height of the patient in cm")
    weight: float = Field(..., description="Weight of the patient in kg")
    ap_hi: int = Field(..., description="Systolic blood pressure")
    ap_lo: int = Field(..., description="Diastolic blood pressure")
    cholesterol: CholesterolLevel = Field(..., description="Cholesterol level")
    gluc: GlucoseLevel = Field(..., description="Glucose level")
    smoke: bool = Field(..., description="Whether the patient smokes")
    alco: bool = Field(..., description="Whether the patient drinks alcohol")
    active: bool = Field(..., description="Whether the patient is physically active")
    cardio: Optional[bool] = Field(None, description="Presence of cardiovascular disease")


class DatasetStatistics(BaseModel):
    """Dataset statistics model."""
    total_records: int = Field(..., description="Total number of records in the dataset")
    cardio_positive: int = Field(..., description="Number of records with cardiovascular disease")
    cardio_negative: int = Field(..., description="Number of records without cardiovascular disease")
    age_range: Dict[str, float] = Field(..., description="Min, max, mean, and median age")
    bmi_range: Dict[str, float] = Field(..., description="Min, max, mean, and median BMI")
    blood_pressure_range: Dict[str, Dict[str, float]] = Field(
        ..., description="Min, max, mean, and median blood pressure (systolic and diastolic)"
    )


class ChartData(BaseModel):
    """Chart data model."""
    chart_type: str = Field(..., description="Type of chart (histogram, scatter, box, etc.)")
    title: str = Field(..., description="Chart title")
    description: str = Field(..., description="Chart description")
    x_label: str = Field(..., description="X-axis label")
    y_label: Optional[str] = Field(None, description="Y-axis label")
    data: List[Dict[str, Union[str, int, float, bool]]] = Field(..., description="Chart data")


class CorrelationAnalysis(BaseModel):
    """Correlation analysis model."""
    correlation_matrix: List[List[float]] = Field(..., description="Correlation matrix")
    feature_names: List[str] = Field(..., description="Feature names")
    top_correlations: List[Dict[str, Union[str, float]]] = Field(
        ..., description="Top correlations with cardiovascular disease"
    )


class Dataset(BaseModel):
    """Dataset model containing all patient records."""
    data: List[Dict[str, Union[str, int, float, bool]]] = Field(..., description="List of patient records")
    total_records: int = Field(..., description="Total number of records in the dataset")

File: ui/src/types/cardio.ts
/**
 * Cardiovascular disease analysis types.
 *
 * This module defines the TypeScript types for cardiovascular disease analysis.
 */

/**
 * Gender enumeration.
 */
export enum Gender {
  FEMALE = 1,
  MALE = 2,
}

/**
 * Cholesterol level enumeration.
 */
export enum CholesterolLevel {
  NORMAL = 1,
  ABOVE_NORMAL = 2,
  WELL_ABOVE_NORMAL = 3,
}

/**
 * Glucose level enumeration.
 */
export enum GlucoseLevel {
  NORMAL = 1,
  ABOVE_NORMAL = 2,
  WELL_ABOVE_NORMAL = 3,
}

/**
 * Patient data model.
 */
export interface PatientData {
  age: number;
  gender: Gender;
  height: number;
  weight: number;
  ap_hi: number;
  ap_lo: number;
  cholesterol: CholesterolLevel;
  gluc: GlucoseLevel;
  smoke: boolean;
  alco: boolean;
  active: boolean;
  cardio?: boolean;
}

/**
 * Dataset statistics model.
 */
export interface DatasetStatistics {
  total_records: number;
  cardio_positive: number;
  cardio_negative: number;
  age_range: {
    min: number;
    max: number;
    mean: number;
    median: number;
  };
  bmi_range: {
    min: number;
    max: number;
    mean: number;
    median: number;
  };
  blood_pressure_range: {
    systolic: {
      min: number;
      max: number;
      mean: number;
      median: number;
    };
    diastolic: {
      min: number;
      max: number;
      mean: number;
      median: number;
    };
  };
}

/**
 * Chart data model.
 */
export interface ChartData<T = Record<string, string | number | boolean>> {
  chart_type: string;
  title: string;
  description: string;
  x_label: string;
  y_label?: string;
  data: Array<T>;
}

export type AgeDistributionChartData = ChartData<{
  age: number;
  num_sick_people: number;
  num_healthy_people: number;
}>;

export type GenderDistributionChartData = ChartData<{
  age: number;
  num_sick_people: number;
  num_healthy_people: number;
  gender: string;
}>;

export type PhysicalActivityChartData = ChartData<{
  active: number;
  num_sick_people: number;
  num_healthy_people: number;
}>;

export type SmokingChartData = ChartData<{
  smoke: number;
  num_sick_people: number;
  num_healthy_people: number;
}>;

export type AlcoholChartData = ChartData<{
  alco: number;
  num_sick_people: number;
  num_healthy_people: number;
}>;

export type CholesterolChartData = ChartData<{
  cholesterol: number;
  num_sick_people: number;
  num_healthy_people: number;
}>;

export type GlucoseChartData = ChartData<{
  gluc: number;
  num_sick_people: number;
  num_healthy_people: number;
}>;

export type BloodPressureCorrelationChartData = ChartData<{
  ap_hi: number;
  ap_lo: number;
  cardio: number;
}>;

export type RiskFactorsRadarChartData = ChartData<{
  factor: string;
  value: number;
}>;

/**
 * Correlation analysis model.
 */
export interface CorrelationAnalysis {
  correlation_matrix: number[][];
  feature_names: string[];
  top_correlations: Array<{
    feature: string;
    correlation: number;
  }>;
}
